### PR TITLE
Add a paddle version of SAC

### PR DESCRIPTION
**Pull request by ShuaibinLi**
### Add files:
mujoco_agent.py, mujoco_model.py, train.py, README.md, ---> benchmark/paddle/SAC/
sac.py ---> parl/algorithms/paddle/

### Change file:
parl/algorithms/paddle/init.py

### CODE REVIEW

File: parl/algorithms/paddle/sac.py [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+import paddle.nn.functional as F
+from copy import deepcopy
+
+__all__ = ['SAC']
+
+
+class SAC(parl.Algorithm):
+    def __init__(self,
+                 model,
+                 target_model,
```
Comment (rical730) [inline parl/algorithms/paddle/sac.py]: please use deepcopy to create target_model

File: parl/algorithms/paddle/sac.py [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+
+        self.gamma = gamma
+        self.tau = tau
+        self.alpha = alpha
+        self.actor_lr = actor_lr
+        self.critic_lr = critic_lr
+
+        self.model = model
+        self.target_model = target_model
+        # self.target_model = deepcopy(self.model)      # Deepcopy can be used with paddle nightly(0.0.0)
```
Comment (rical730) [inline parl/algorithms/paddle/sac.py]: please use deepcopy to create target_model

File: benchmark/paddle/SAC/train.py [hunk 3]
```diff
@@ -0,0 +1,162 @@
...
+    env = gym.make(args.env)
+    env.seed(args.seed)
+    env = ActionMappingWrapper(env)
+
+    obs_dim = env.observation_space.shape[0]
+    action_dim = env.action_space.shape[0]
+
+    # Initialize model, algorithm, agent, replay_memory
+    model = MujocoModel(obs_dim, action_dim)
+    target_model = MujocoModel(obs_dim, action_dim)
```
Comment (rical730) [inline benchmark/paddle/SAC/train.py]: please use deepcopy to create target_model

File: benchmark/paddle/SAC/README.md [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+<img src=".benchmark/SAC_results.png" width = "800" height ="400" alt="SAC_results"/>
+
++ Each experiment was run three times with different seeds
+
+## How to use
+### Dependencies:
++ python3.5+
++ [parl==1.4.1](https://github.com/PaddlePaddle/PARL)
++ gym==0.9.1
++ [paddlepaddle==2.0.0rc1](https://github.com/PaddlePaddle/Paddle)
```
Comment (rical730) [inline benchmark/paddle/SAC/README.md]: Remind：please update paddle version after next version later than rc1 published

File: benchmark/paddle/SAC/README.md [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+Based on PARL, the SAC algorithm of deep reinforcement learning has been reproduced, reaching the same level of indicators as the paper in Mujoco benchmarks.
+
+> Paper: SAC in [Soft Actor-Critic: Off-Policy Maximum Entropy Deep Reinforcement Learning with a Stochastic Actor](https://arxiv.org/abs/1801.01290)
+
+### Mujoco games introduction
+Please see [here](https://github.com/openai/mujoco-py) to know more about Mujoco games.
+
+### Benchmark result
+
+<img src=".benchmark/SAC_results.png" width = "800" height ="400" alt="SAC_results"/>
```
Comment (TomorrowIsAnOtherDay) [inline benchmark/paddle/SAC/README.md:11]: Can you add the figures for other environments? 
There is only one environment here.

File: benchmark/paddle/SAC/README.md (hunk 4 above)
Comment (ShuaibinLi) [reply benchmark/paddle/SAC/README.md]: done

File: benchmark/paddle/SAC/README.md (hunk 5 above)
Comment (ShuaibinLi) [reply benchmark/paddle/SAC/README.md:11]: done

File: benchmark/paddle/SAC/README.md [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+
++ Each experiment was run three times with different seeds
+
+## How to use
+### Dependencies:
++ python3.5+
++ [parl==1.4.1](https://github.com/PaddlePaddle/PARL)
```
Comment (rical730) [inline benchmark/paddle/SAC/README.md]: parl>=1.4.2

File: benchmark/paddle/SAC/README.md [hunk 7]
```diff
@@ -0,0 +1,31 @@
...
++ gym==0.9.1
++ [paddlepaddle==2.0.0](https://github.com/PaddlePaddle/Paddle)
```
Comment (rical730) [inline benchmark/paddle/SAC/README.md]: paddlepaddle>=2.0.0

File: benchmark/paddle/SAC/train.py [hunk 8]
```diff
@@ -0,0 +1,160 @@
...
+
+if __name__ == "__main__":
+    parser = argparse.ArgumentParser()
+    parser.add_argument(
+        "--env", default="HalfCheetah-v1", help='Mujoco gym environment name')
+    parser.add_argument(
+        "--seed",
+        default=0,
+        type=int,
+        help='Sets Gym, PyTorch and Numpy seeds')
```
Comment (rical730) [inline benchmark/paddle/SAC/train.py]: remove PyTorch and Numpy

File: parl/algorithms/paddle/sac.py [hunk 9]
```diff
@@ -0,0 +1,121 @@
...
+                 tau=None,
+                 alpha=None,
+                 actor_lr=None,
+                 critic_lr=None):
+        """ SAC algorithm
+            Args:
+                model(parl.Model): forward network of actor and critic.
+                gamma(float): discounted factor for reward computation
+                tau (float): decay coefficient when updating the weights of self.target_model with self.model
+                alpha (float): Temperature parameter determines the relative importance of the entropy against the reward
```
Comment (rical730) [inline parl/algorithms/paddle/sac.py]: temperature

File: benchmark/paddle/SAC/train.py [hunk 10]
```diff
@@ -0,0 +1,160 @@
+#   Copyright (c) 2020 PaddlePaddle Authors. All Rights Reserved.
```
Comment (rical730) [inline benchmark/paddle/SAC/train.py]: 2020 --> 2021，please update all python files.

File: benchmark/paddle/SAC/train.py (hunk 8 above)
Comment (rical730) [reply benchmark/paddle/SAC/train.py]: seeds --> seed

File: benchmark/paddle/SAC/README.md [hunk 11]
```diff
@@ -0,0 +1,31 @@
...
+Please see [here](https://github.com/openai/mujoco-py) to know more about Mujoco games.
+
+### Benchmark result
+
+<img src=".benchmark/SAC_results.png" width = "800" height ="400" alt="SAC_results"/>
+
++ Each experiment was run three times with different seeds
+
+## How to use
+### Dependencies:
```
Comment (rical730) [inline benchmark/paddle/SAC/README.md:16]: Dependencies recommended order:
python
parl
paddle
gym
mujoco

File: benchmark/paddle/SAC/train.py [hunk 12]
```diff
@@ -0,0 +1,156 @@
...
+ACTOR_LR = 3e-4
+CRITIC_LR = 3e-4
+
+
+# Run episode for training
+def run_train_episode(agent, env, rpm):
+    action_dim = env.action_space.shape[0]
+    obs = env.reset()
+    done = False
+    episode_reward, episode_steps = 0, 0
```
Comment (rical730) [inline benchmark/paddle/SAC/train.py]: episode_reward = 0
episode_steps = 0

File: parl/algorithms/paddle/sac.py [hunk 13]
```diff
@@ -0,0 +1,121 @@
...
+        self.tau = tau
+        self.alpha = alpha
+        self.actor_lr = actor_lr
+        self.critic_lr = critic_lr
+
+        self.model = model
+        self.target_model = deepcopy(self.model)
+        self.actor_optimizer = paddle.optimizer.Adam(
+            learning_rate=actor_lr,
+            parameters=self.model.actor_model.parameters())
```
Comment (rical730) [inline parl/algorithms/paddle/sac.py]: self.model.get_actor_params()

File: parl/algorithms/paddle/sac.py [hunk 14]
```diff
@@ -0,0 +1,121 @@
...
+        self.critic_optimizer = paddle.optimizer.Adam(
+            learning_rate=critic_lr,
+            parameters=self.model.critic_model.parameters())
```
Comment (rical730) [inline parl/algorithms/paddle/sac.py]: self.model.get_critic_params()

Comment (ShuaibinLi) [reply parl/algorithms/paddle/sac.py]: done

**Review by TomorrowIsAnOtherDay — DISMISSED** [review]
LGTM.

**Review by TomorrowIsAnOtherDay — APPROVED** [review]
